README.rst
==================================
Grove GPS
==================================

API for the Seeed Studio Grove GPS.

* GitHub repo: https://github.com/Paolo-dono/Grove_GPS
* Free software: MIT License


docs/history.txt
2020-10-13 19:43:13.837867
19 Loots Road Blairgowrie, Randburg
2020-10-13 19:52:11.478735
19 Loots Road Blairgowrie, Randburg
2020-10-13 20:20:52.969457
17 Loots Road Blairgowrie, Randburg
2020-10-13 20:21:37.205097
15 Loots Road Blairgowrie, Randburg
2020-10-13 20:22:13.850905
13 Loots Road Blairgowrie, Randburg
2020-10-13 20:24:17.735032
11 Loots Road Blairgowrie, Randburg
2020-10-13 20:24:55.804660
9 Loots Road Blairgowrie, Randburg
2020-10-20 14:02:34.178055
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 14:04:47.059459
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 14:07:03.250895
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 14:15:15.530264
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 17:40:51.478875
17 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 18:19:30.839040
22 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 18:22:14.997994
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 19:29:01.308278
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 20:00:50.561547
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 20:05:50.375595
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-20 21:49:09.428676
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 17:37:42.727394
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 17:54:41.299521
1 Loots Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:21:43.436661
70 Conrad Dr, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:21:46.240308
64 Conrad Dr, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:24:47.596888
1 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:25:21.178946
3 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:25:32.416150
5 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:25:54.758731
2 Eileen Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:26:06.073859
4 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:26:39.765222
7 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:26:50.963620
6 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:27:02.310324
8 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:27:13.946155
10 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:27:25.314233
15 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:27:36.639041
14 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:27:47.832602
16 Geneva Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:28:12.129021
Bramfischer Ave & Grove Rd, Blairgowrie, Randburg, 2194, South Africa
2020-10-21 18:28:35.146366
11 Grove St, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:28:53.637847
3rd Floor, Oak Place, West Wing, Oak Ave, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:29:06.364633
19 Grove St, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:29:09.908270
21A Grove St, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:29:18.322430
23 Grove St, Ferndale, Randburg, 2160, South Africa
2020-10-21 18:29:22.065436
25 Vale Ave, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:29:36.515326
29 Grove St, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:33:55.049118
37 Grove St, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:34:36.067270
220 Main Ave, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:34:57.189644
206 Main Ave, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:35:12.171411
61 Cross Street, President Ridge, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:35:26.299368
43 Cross St, President Ridge, Randburg, 2194, South Africa
2020-10-21 18:35:29.084375
53 Cross St, President Ridge, Randburg, 2194, South Africa
2020-10-21 18:35:37.479986
53 Cross St, President Ridge, Randburg, 2194, South Africa
2020-10-21 18:36:02.750496
43 Cross St, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:37:32.559516
239 Vale Ave, Ferndale, Randburg, 2194, South Africa
2020-10-21 18:37:55.999616
Brightwater Commons, 50 Vale Ave, Ferndale, Randburg, 2194, South Africa
2020-10-21 19:36:17.798852
19 Loots Rd, Blairgowrie, Randburg, 2194, South Africa


docs/test.py
from location_calcs import location_calcs

g = location_calcs()
print(g.locationAtTime("2020-10-14 20:22:14.850905"))

grove_gps.py
#The GPS module used is a Grove GPS module http://www.seeedstudio.com/depot/Grove-GPS-p-959.html

import serial, time
    
ser = serial.Serial('/dev/ttyAMA0',  9600, timeout = 0)
ser.flush()

def cleanstr(in_str):
    out_str = "".join([c for c in in_str if c in "0123456789.-" ])
    if len(out_str)==0:
        out_str = "-1"
    return out_str

def safefloat(in_str):
    try:
        out_str = float(in_str)
    except ValueError:
        out_str = -1.0
    return out_str

# Convert to decimal degrees
def decimal_degrees(raw_degrees):
    try:
        degrees = float(raw_degrees) // 100
        d = float(raw_degrees) % 100 / 60
        return degrees + d
    except: 
        return raw_degrees

class GPS:
    inp = []
    inp2 = []
    GGA = []
    RMC = []
    values = []
    
    def __init__(self):
        self.refresh()
    
    #Read data from the GPS
    def refresh(self): 
        while True:
            line = ser.readline()
            GPS.inp = line.decode('ISO-8859-1')
            #print(GPS.inp + "\n")
            if GPS.inp[0:6] =='$GPGGA': # GGA data for latitude, longitude, satellites, altitude, UTC position
                GPS.GGA = GPS.inp.split(",")
                if len(GPS.GGA) >= 10:
                    break
            time.sleep(0.1)     #needed by the cmd in order not to crash
        while True:
            GPS.inp2 = ser.readline().decode()
            if GPS.inp2[0:6] =='$GPRMC': # RMC data to get velocity
                GPS.RMC = GPS.inp2.split(",")
                if len(GPS.RMC) >= 8:
                    break
            time.sleep(0.1)
        
        #initialize values obtained from the GPS device
        
        if GPS.GGA[1] == '':
            ti = ""
        else:
            ti = GPS.GGA[1].split(".")[0]
            ti = ti[0:2] + ":" + ti[2:4] + ":" + ti[4:]  # convert to standard time format
        
        if GPS.GGA[2] == '':  # latitude. Technically a float
            lat =-1.0
        else:
            lat = decimal_degrees(safefloat(cleanstr(GPS.GGA[2])))
        
        if GPS.GGA[3] == '':  # this should be either N or S
            lat_ns = ""
        else:
            lat_ns=str(GPS.GGA[3])
        if lat_ns == "S":
            lat = -lat
            
        if  GPS.GGA[4] == '':  # longitude. Technically a float
            long = -1.0
        else:
            long = decimal_degrees(safefloat(cleanstr(GPS.GGA[4])))
        
        if  GPS.GGA[5] == '': # this should be either W or E
            long_ew = ""
        else:
            long_ew = str(GPS.GGA[5])
        if long_ew == "W":
            long = -long
        
        if GPS.GGA[7] == '': # number of satellites
            sats = 0
        else:
            sats = int(cleanstr(GPS.GGA[7]))
        
        if  GPS.GGA[9] == '': # altitude
            alt = -1.0
        else:
            alt = GPS.GGA[9]
        
        if GPS.RMC[7] == '': # speed
            spd = 0.0
        else:
            spd = 1.852*safefloat(GPS.RMC[7]) # conversion from knots to km/h
        
        GPS.values = [lat,lat_ns,long,long_ew,sats,alt,spd,ti]
    
    # Accessor methods for all the desired GPS values
    def getLatitude(self):
        return GPS.values[0]
    def getNS(self):
        return GPS.values[1]
    def getLongitude(self):
        return GPS.values[2]
    def getEW(self):
        return GPS.values[3]
    def getSatellites(self):
        return GPS.values[4]
    def getAltitude(self):
        return GPS.values[5]
    def getSpeed(self):
        return GPS.values[6]
    def getUTCPosition(self):
        return GPS.values[7]
        
if __name__ == "__main__":
    gps = GPS()
    
    # if all the folowing outputs have values,
    # it show the GPS module was able to connect to the satellites properly
    print("Latitude: " + str(gps.getLatitude()))
    print("Longitude: " + str(gps.getLongitude()))
    print("Number of satellites: " + str(gps.getSatellites()))
    print("UTC position: " + gps.getUTCPosition())
    
    
    

location_calcs.py
from grove_gps import GPS
import time, datetime
import json
import math
import urllib.request

class location_calcs:
    g = None
    refresh = 30
    
    def __init__(self): # initialises the global variables
        global g
        global refresh
        
        g = GPS()
        refresh = 30
    
    def getLogRefreshRate(self): # returns the rate at which new locations are added 
        global refresh           # to the history log of locations visited
        return refresh
    
    def setLogRefreshRate(self, new_refresh): # sets a new refresh rate for the history log
        global refresh
        refresh = new_refresh
    
    def getCurrentAddress(self): # performs reverse geocoding to return the current street address
        global g                 # from the coordinates read from the GPS module
        
        g.refresh()
        api_key = "AIzaSyBye-KERYIXDd24hk0C19mSMXdS7TmCX3M"
        coordinates = str(g.getLatitude()) + "," + str(g.getLongitude())
        link = "https://maps.googleapis.com/maps/api/geocode/json?latlng={}&key={}".format(coordinates, api_key)
        response = urllib.request.urlopen(link).read()
        address = json.loads(response)["results"][0]["formatted_address"]
        return address
    
    def getAddressFromCoordinates(self, lat, long): # returns the street address given cooordinates as arguments
        api_key = "AIzaSyBye-KERYIXDd24hk0C19mSMXdS7TmCX3M"
        coordinates = str(lat) + "," + str(long)
        link = "https://maps.googleapis.com/maps/api/geocode/json?latlng={}&key={}".format(coordinates, api_key)
        response = urllib.request.urlopen(link).read()
        address = json.loads(response)["results"][0]["formatted_address"]
        return address
    
    def getCoordinatesFromAddress(self, address): # performs geocoding to return coordinates given a street address
        api_key = "AIzaSyBye-KERYIXDd24hk0C19mSMXdS7TmCX3M"
        address = address.replace(" ", "+")
        link = "https://maps.googleapis.com/maps/api/geocode/json?address={}&key={}".format(address, api_key)
        response = urllib.request.urlopen(link).read()
        data = json.loads(response)
        latitude = data["results"][0]["geometry"]["location"]["lat"]
        longitude = data["results"][0]["geometry"]["location"]["lng"]
        return [latitude, longitude]
    
    def distance(self, lat1, lng1, lat2, lng2): # finds the distance between two sets of coordinates
        degtorad = math.pi/180
        dLat = (lat1-lat2)*degtorad
        dLng = (lng1-lng2)*degtorad
        a = pow(math.sin(dLat/2), 2) + math.cos(lat1*degtorad)*math.cos(lat2*degtorad)*pow(math.sin(dLng/2), 2)
        b = 2*math.atan2(math.sqrt(a), math.sqrt(1-a))
        return 6371*b
    
    def distanceToHome(self): # finds the distance to the home address that was set
        f = open("home.txt", "r")
        home = f.readlines()[0]
        f.close()
        lat, lng = g.getCoordinatesFromAddress(home)
        g.refresh()
        lat2 = g.getLatitude()
        lng2 = g.getLongitude()
        return self.distance(lat, lng, lat2, lng2)
    
    def setHomeAddress(self, address): # sets home address
        f = open("home.txt", "w")
        f.write(home)
        f.close()
    
    def getHomeAddress(self): # returns home address
        f = open("home.txt", "r")
        a = f.readlines()
        return a[0]
    
    def logLocation(self):
        global refresh
        
        time_start = time.time()
        dt = datetime.datetime.now()
        address1 = self.getCurrentAddress()
        h = open("history.txt", "a")
        h.write(str(dt) + "\n")
        h.write(address1 + "\n")
        h.close()
        while True:
            address2 = self.getCurrentAddress()
            print(address2+"\n")
            if address1 != address2:
                time_elapsed = time.time() - time_start
                print(time_elapsed + "\n")
                if time_elapsed > 600:
                    favourites = open("favourites.txt")
                    fav = json.load(favourites)
                    favourites.close()
                    if address2 in fav:
                        fav[address2] += 1
                    else:
                        fav[address2] = 1
                    new_favourites = open("favourites.txt", "w")
                    json.dump(fav, new_favourites)
                    new_favourites.close()
                dt = datetime.datetime.now()
                f = open("history.txt", "a")
                f.write(str(dt) + "\n")
                f.write(address2 + "\n")
                f.close()
                address1 = address2
                time_start = time.time()
            time.sleep(refresh)
    
    def getFavouriteLocations(self):
        favourites = open("favourites.txt")
        fav = json.load(favourites)
        favourites.close()
        top = []
        for i in range(5):
            highest = 0
            place = ""
            for j in fav:
                if fav[j] > highest:
                    highest = fav[j]
                    place = j
            fav.pop(place)
            top.append(place)
        return top
    
    def locationAtTime(self, dt):
        f = open("history.txt", "r")
        data = f.readlines()
        f.close()
        start = 0
        stop = len(data)
        found = False
        address = ""
        while found == False:
            pos = (stop - start)//2 + start
            if pos%2 == 1:
                pos-=1
            if dt < data[pos][:19]:
                stop = pos
            else:
                start = pos
            if (stop - start) == 2:
                found = True
                address = data[pos+1]
        return address
    
    def timesAtLocation(self, address):
        favourites = open("favourites.txt")
        fav = json.load(favourites)
        favourites.close()
        return fav[address]

if __name__ == "__main__":
    gps = location_calcs()
    #gps.logLocation()
#     print(gps.getCoordinatesFromAddress("19 Loots Road Blairgowrie, Randburg"))
#     print(gps.distance(-26.103062, 28.003815, -26.116043, 28.001292))
#     print(gps.locationAtTime("2020-10-14 20:22:14.850905"))
#     print(gps.timesAtLocation("address3"))
